src/BalanceLaws/interface.jl

```diff
@@ -316,18 +262,11 @@ Compute reverse indefinite integral along stack.
 function reverse_indefinite_stack_integral! end
 
 # Internal methods
-number_state_conservative(m::BalanceLaw, FT) =
-    varsize(vars_state_conservative(m, FT))
-number_state_auxiliary(m::BalanceLaw, FT) = varsize(vars_state_auxiliary(m, FT))
-number_state_gradient(m::BalanceLaw, FT) = varsize(vars_state_gradient(m, FT))
-number_state_gradient_flux(m::BalanceLaw, FT) =
-    varsize(vars_state_gradient_flux(m, FT))
-num_gradient_laplacian(m::BalanceLaw, FT) =
-    varsize(vars_gradient_laplacian(m, FT))
-num_hyperdiffusive(m::BalanceLaw, FT) = varsize(vars_hyperdiffusive(m, FT))
-num_integrals(m::BalanceLaw, FT) = varsize(vars_integrals(m, FT))
-num_reverse_integrals(m::BalanceLaw, FT) =
-    varsize(vars_reverse_integrals(m, FT))
+number_states(m::BalanceLaw, st::AbstractStateType, FT) =
+    varsize(vars_state(m, st, FT))
+
+rank_multiplier(st::AbstractStateType) = 1
+rank_multiplier(st::GradientLaplacian) = 3
```

Comment:
We might want to expand this to include `::Gradient` and maybe `::GradientFlux`

Reply:
`Gradient` and `GradientFlux` are subtypes of `AbstractStateType`, so these methods provide fallbacks. Maybe I’m misunderstanding. Can you give an example?

Reply:
Or maybe this is just a bad name. I think I wasn't sure what to call this function, actually. It's only used to account for this factor of 3 [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/DGMethods/create_states.jl#L139). The other methods (for [Hyperdiffusive](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/DGMethods/create_states.jl#L156), [GradientFlux](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/DGMethods/create_states.jl#L106), [Auxiliary](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/DGMethods/create_states.jl#L49), and [Conservative](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/DGMethods/create_states.jl#L19)) all have coefficients 1. Is this a good name? or do you have a better idea for a name here?

Reply:
Is the rank multiplier the change from one component scalars to three component gradients? I just meant to include more things in the 3 category.

Reply:
> Is the rank multiplier the change from one component scalars to three component gradients?

🤷‍♂️ I was just generalizing `create_conservative_state`, `create_auxiliary_state`, etc.. Maybe @jkozdon or @lcw can comment? Using

`rank_multiplier(::Gradient) = 3`
`rank_multiplier(::GradientFlux) = 3`

would be behavior changing, and I'm not sure it would work / be correct.

Reply:
I personally don't like this. Can we just leave the 3 where its needed. This is going to obfuscate what's happening.

Reply:
> I personally don't like this. Can we just leave the 3 where its needed. This is going to obfuscate what's happening.

So, specialize on `create_state(:: GradientLaplacian)` (have a method for `AbstractStateType` and `GradientLaplacian` methods) and hard-code the 3 in the `GradientLaplacian` method?

Reply:
Sure.

It just seems a bit opaque to have this radom `rank_multiplier` function elsewhere.

Reply:
`rank_multiplier` is only needed for `create_state`, we could just move it right above `create_state` instead